source/common/http/conn_pool_base.h

```diff
@@ -70,61 +74,51 @@ class ActiveClient : public LinkedObject<ActiveClient>,
   bool timed_out_{false};
 };
 
-using ActiveClientPtr = std::unique_ptr<ActiveClient>;
-
+// PendingRequest is the base class for a connection which has been created but not yet established.
 class PendingRequest : public LinkedObject<PendingRequest>, public ConnectionPool::Cancellable {
 public:
-  PendingRequest(ConnPoolImplBase& parent, ResponseDecoder& decoder,
-                 ConnectionPool::Callbacks& callbacks);
+  PendingRequest(ConnPoolImplBase& parent);
   ~PendingRequest() override;
 
   // ConnectionPool::Cancellable
   void cancel(Envoy::ConnectionPool::CancelPolicy policy) override;
 
+  // The context here returns a pointer to whatever context is provided with newStream(),
+  // which will be passed back to the parent in onPoolReady or onPoolFailure.
+  virtual void* context() PURE;
+
   ConnPoolImplBase& parent_;
-  ResponseDecoder& decoder_;
-  ConnectionPool::Callbacks& callbacks_;
 };
 
 using PendingRequestPtr = std::unique_ptr<PendingRequest>;
 
+using ActiveClientPtr = std::unique_ptr<ActiveClient>;
+
 // Base class that handles request queueing logic shared between connection pool implementations.
-class ConnPoolImplBase : public ConnectionPool::Instance,
-                         protected Logger::Loggable<Logger::Id::pool> {
+class ConnPoolImplBase : protected Logger::Loggable<Logger::Id::pool>,
+                         public virtual ConnectionPool::Instance {
```

Comment:
I think you can delete this line, so it's just `class ConnPoolImplBase : protected Logger::Loggable<Logger::Id::pool> {` because this class no longer overrides any of the methods on that interface. That's what made it seem like the virtual inheritance in the other location wasn't needed.

Reply:
oh, gotcha. Sorry for missing what you were saying previously!